File: Gemi3/pyCAVI/_update_r.py
import numpy as np
import sys
from Gemi3.pyCAVI.util import find_s_keys,find_r_keys

def r_loop(self):
    if self.verbose:
        print("Updating r values",file=sys.stdout)
    for r_triple in self.hash_r.keys():
        indices=collect_all_keys(self,r_triple)
        numerator_sum=np.sum(np.asarray(list(map(lambda x: r_numerator(self,x),indices))))
        denominator_sum=np.sum(np.asarray(list(map(lambda x: r_denominator(self,x),indices))))
        new_r=((self.prio_tau_r*self.prio_mu_r)+numerator_sum)/(self.prio_tau_r+denominator_sum)
        if self.debug: print(f"Old r {self.r[self.hash_r[r_triple]]}.", sys.stdout)
        self.r[self.hash_r[r_triple]]= new_r
        if self.debug: print(f"New r {self.r[self.hash_r[r_triple]]}.", sys.stdout)

def rr_loop(self):
    if self.verbose:
        print("Updating rr values",file=sys.stdout)
    for r_triple in self.hash_r.keys():
        indices=collect_all_keys(self,r_triple)
        numerator_sum=np.sum(np.asarray(list(map(lambda x: r_numerator(self,x),indices))))
        denominator_sum=np.sum(np.asarray(list(map(lambda x: r_denominator(self,x),indices))))
        new_rr=((self.prio_tau_rr*self.prio_mu_rr)+numerator_sum)/(self.prio_tau_rr+denominator_sum)
        if self.debug: print(f"Old rr {self.rr[self.hash_r[r_triple]]}.", sys.stdout)
        self.rr[self.hash_r[r_triple]]= new_rr
        if self.debug: print(f"New rr {self.rr[self.hash_r[r_triple]]}.", sys.stdout)

def r_numerator(self,key_dict):
    numerator = (self.tau[self.hash_t[key_dict["t"]]] * self.s[self.hash_s[key_dict["s"]]]) * \
                (self.D[self.hash_d[key_dict["t"]]] - \
                 (self.x[self.hash_x[key_dict["hj"]]] * self.y[self.hash_y[key_dict["hl"]]]) - \
                 (self.x[self.hash_x[key_dict["fk"]]] * self.y[self.hash_y[key_dict["fl"]]]) - \
                 (self.x[self.hash_x[key_dict["gi"]]] * self.y[self.hash_y[key_dict["gl"]]]))
    return numerator

def r_denominator(self,key_dict):
    denominator = self.tau[self.hash_t[key_dict["t"]]] * self.ss[self.hash_s[key_dict["s"]]]
    return denominator

def collect_all_keys(self,k_triple):
    #only get keys from tau and split them accordingly
    tau_keys = [x for x in self.hash_t.keys() if k_triple in x]
    out=[]
    for k in tau_keys:
        gi,hj,fk,l = k.split("_")
        gl = self.guide_to_gene(gi)
        hl = self.guide_to_gene(hj)
        fl = self.guide_to_gene(fk)
        s = find_s_keys(self,gl,hl,fl,l)
        r = find_r_keys(self,gi,hj,fk)
        out.append({"t":k,"gi":gi,"hj":hj,"fk":fk,"l":l,
                    "gl":"_".join([gl,l]),
                    "hl":"_".join([hl,l]),
                    "fl":"_".join([hl,l]),
                    "r":r,
                    "s":s})
    return out

File: src/CAVI/update_r.rs
use super::CAVI;
use std::collections::HashMap;
use itertools::Itertools;
use rayon::prelude::*;

impl<'a> CAVI<'a>{
    pub fn update_r(&self) -> Result<(),Box<dyn std::error::Error>>{
        if self.verbose {
            println! {"Updating r values"}
        }
        for gl in self.lookup.read().unwrap().get("r").unwrap().keys(){
            let all_numerators :Vec<f64> = self.lookup.read().unwrap().get("r").unwrap().get(gl).unwrap().par_iter().map(|element| r_numerator(self,element)).collect();
            let all_denominators :Vec<f64> = self.lookup.read().unwrap().get("r").unwrap().get(gl).unwrap().par_iter().map(|element| r_denominator(self,element)).collect();
            let new_r : f64 = ((self.prio_tau_r * self.prio_mu_r) + all_numerators.par_iter().sum::<f64>())/(self.prio_tau_r+all_denominators.par_iter().sum::<f64>());
            if self.debug {println!("Old r {}.",*self.all_r.read().unwrap().get(*self.hash_r.get(*gl).unwrap()).unwrap())}
            {*self.all_r.write().unwrap().get_mut(*self.hash_r.get(*gl).unwrap()).unwrap()= new_r}
            if self.debug {println!("New r {}.",*self.all_r.read().unwrap().get(*self.hash_r.get(*gl).unwrap()).unwrap())}
        }
        return Ok(());
    }

    pub fn update_rr(&self) -> Result<(),Box<dyn std::error::Error>>{
        if self.verbose{
            println!{"Updating rr values"}
        }
        for gl in self.lookup.read().unwrap().get("r").unwrap().keys(){
            let all_numerators :Vec<f64> = self.lookup.read().unwrap().get("r").unwrap().get(gl).unwrap().par_iter().map(|element| r_numerator(self,element)).collect();
            let all_denominators :Vec<f64> = self.lookup.read().unwrap().get("r").unwrap().get(gl).unwrap().par_iter().map(|element| r_denominator(self,element)).collect();
            let new_rr : f64 = ((self.prio_tau_rr * self.prio_mu_rr) + all_numerators.par_iter().sum::<f64>())/(self.prio_tau_rr+all_denominators.par_iter().sum::<f64>());
            if self.debug {println!("Old rr {}.",*self.all_rr.read().unwrap().get(*self.hash_r.get(*gl).unwrap()).unwrap())}
            {*self.all_rr.write().unwrap().get_mut(*self.hash_r.get(*gl).unwrap()).unwrap()= new_rr}
            if self.debug {println!("New rr {}.",*self.all_rr.read().unwrap().get(*self.hash_r.get(*gl).unwrap()).unwrap())}
        }
        return Ok(());
    }

}

fn r_numerator(obj : &CAVI,key_dict : &HashMap<&str,&str>) -> f64 {
    let numerator : f64 = (*obj.all_tau.read().unwrap().get(*obj.hash_td.get(*key_dict.get("t").unwrap()).unwrap()).unwrap() *
        *obj.all_s.read().unwrap().get(*obj.hash_s.get(*key_dict.get("s").unwrap()).unwrap()).unwrap()) *
        (*obj.all_D.read().unwrap().get(*obj.hash_td.get(*key_dict.get("t").unwrap()).unwrap()).unwrap() -
        *obj.all_x.read().unwrap().get(*obj.hash_x.get(*key_dict.get("hj").unwrap()).unwrap()).unwrap() * *obj.all_y.read().unwrap().get(*obj.hash_y.get(*key_dict.get("hl").unwrap()).unwrap()).unwrap() -
        *obj.all_x.read().unwrap().get(*obj.hash_x.get(*key_dict.get("fk").unwrap()).unwrap()).unwrap() * *obj.all_y.read().unwrap().get(*obj.hash_y.get(*key_dict.get("fl").unwrap()).unwrap()).unwrap() -
        *obj.all_x.read().unwrap().get(*obj.hash_x.get(*key_dict.get("gi").unwrap()).unwrap()).unwrap() * *obj.all_y.read().unwrap().get(*obj.hash_y.get(*key_dict.get("gl").unwrap()).unwrap()).unwrap());
    return numerator;
    }


fn r_denominator(obj : &CAVI,key_dict : &HashMap<&str,&str>) -> f64 {
    let denominator : f64 = *obj.all_tau.read().unwrap().get(*obj.hash_td.get(*key_dict.get("t").unwrap()).unwrap()).unwrap() *
        *obj.all_ss.read().unwrap().get(*obj.hash_s.get(*key_dict.get("s").unwrap()).unwrap()).unwrap();
    return denominator;
    }